### PR TITLE
Convert CRWA seed load script to Typescript

### PR DESCRIPTION
**Pull request by dthyresson**
This PR converts the `seed.js` Prisma load script to Typescript according to the instructions found: https://www.prisma.io/docs/guides/database/seed-database#seeding-your-database-with-typescript

It:

* renames file
* uses import rather than require
* adds dev dependencies for ts-node and @types/node
* adds additional --compiler-options to the script that Prisma uses to run the seed script

These changes do work:

image

But, I'm not certain that if:

* dev dependencies are in the right package
* if compiler options should be in tsconfig?
* should CRWA support a seed.js and seed.ts (and if so, how to install the seed file/add dependencies based on if project is typescript or not)

### CODE REVIEW

**Comment by dac09** [comment]
Nice DT, yeah I'm wondering the same about the dependencies. Perhaps what I would do here is have an additional step in CRWA if its a typescript project that adds tsnode as dep and the script

**Comment by dthyresson** [comment]
One idea I had that I confirmed seems like it could work is to add the `api/db` path to the ts-to-js conversion script:

```
// convertTsProjectToJs in internal/devtools
export const typeScriptSourceFiles = (cwd: string) => {
  // TODO: When sides are expanded read the `api` and `web` string instead
  // of hard-coding them.
  return glob.sync('{api,web}/{src,db}/**/*.{ts,tsx}', {
    cwd,
  })
}
```

This *does* convert seed.ts to seed.js but seeding then complains:

```terminal
Running Prisma CLI:
yarn prisma db seed --preview-feature --schema "/Users/dthyresson/redwood-tests/seedtest9/api/db/schema.prisma" 

Prisma schema loaded from db/schema.prisma
Running node "db/seed.js" ...
/Users/dthyresson/redwood-tests/seedtest9/api/db/seed.js:2
import { PrismaClient } from '@prisma/client'
^^^^^^

SyntaxError: Cannot use import statement outside a module
    at wrapSafe (internal/modules/cjs/loader.js:1172:16)
    at Module._compile (internal/modules/cjs/loader.js:1220:27)
    at Object.Module._extensions..js (internal/modules/cjs/loader.js:1277:10)
    at Module.load (internal/modules/cjs/loader.js:1105:32)
    at Function.Module._load (internal/modules/cjs/loader.js:967:14)
    at Function.executeUserEntryPoint [as runMain] (internal/modules/run_main.js:60:12)
    at internal/main/run_main_module.js:17:47
Error: Command failed with exit code 1: node "db/seed.js"
error Command failed with exit code 1.
```

So, if can sort that out, then don't need any conditional logic, the TS to JS conversion will handle.

**Comment by dac09** [comment]
That import thingy happens, because the typescript file isn't preprocessed (not babel'ed). I guess internally prisma cli uses tsnode, when it sees it configured on the project.

**Comment by dthyresson** [comment]
> because the typescript file isn't preprocessed (not babel'ed).

Is there a way for me to fix that?

**Comment by dac09** [comment]
Change would have to be on the prisma end I'm afraid, because they're ones invoking the seed script, not us.

**Comment by dthyresson** [comment]
> Change would have to be on the prisma end I'm afraid, because they're ones invoking the seed script, not us.

Could I pre-process `seed.ts` in the `ts-to-js` script to convert?

File: packages/create-redwood-app/src/create-redwood-app.js [hunk 1]
```diff
@@ -129,6 +129,27 @@ const installNodeModulesTasks = ({ newAppDir }) => {
         })
       },
     },
+    {
+      title: 'Updating Prisma seed file imports',
```
Comment (dthyresson) [inline packages/create-redwood-app/src/create-redwood-app.js:133]: This task replaces the first 2 lines of the seed.ts with `require`s and renames the seed.ts to seed.js.

Perhaps instead can use babel?

**Comment by dthyresson** [comment]
@peterp Please see the `title: 'Updating Prisma seed file imports'` task.

This simply replaces the imports for the JS version of the seed file.

Ideally, this would be transpiled to JS instead I imagine but I did not see a way for `convertTsProjectToJs` to add that api/db path.

**Comment by dthyresson** [comment]
@dac09 and I paired and decided that for simplicity, clarity and future use (ie, to have specific seed examples for Typescript using actual Prisma types) it would be better to keep two separate ts and js template files and create that file as `api/db/seed` with the proper expected extension in CRWA.

**Review by dac09 — APPROVED** [review]
This is the simplest solution for now, and I like it! We can iterate on this as our requirements change, or as prisma updates their workflow for seeding dbs

**Comment by dac09** [comment]
Something else I would mention in the docs for this @dthyresson - if you needed more powerful seeding functions that access your services, you can always generate a script with `yarn rw g script seed` and `yarn rw exec seed`

**Review by peterp — CHANGES_REQUESTED** [review]
I actually don't think this is the correct approach because using `seed.ts` in this way will diverge the way we handle transpilation into two paths: the Redwood way (babel), and then a completely separate way (ts-node), but without any of our transformations. Sorry @dthyresson!

This may be problematic if people import other files in `src` and expect it to behave the same. I think an approach could be to overwrite the `prisma db seed` command and wire it up with our own babel transpilation instead of using `ts-node`

**Comment by dthyresson** [comment]
> I think an approach could be to overwrite the `prisma db seed` command and wire it up with our own babel transpilation instead of using `ts-node`

@peterp Prisma seed is still a preview feature. Should we get their input before changing their command -- which does require ts-node as per https://www.prisma.io/docs/guides/database/seed-database#requirements-for-seeding-with-typescript-or-javascript

**Comment by peterp** [comment]
@dthyresson I think that's a good idea - we could suggest to them that we want to use our own transpilation - perhaps there's a way that they could allow us to specify a command, or a pre-transpile function.

**Comment by dthyresson** [comment]
Assigning to @peterp per standup 6/28 as has ideas to transpile vs swap out.

**Comment by peterp** [comment]
Ok, I'm going to overwrite the `yarn rw prisma db seed` command so that it uses our babel scripts instead of Prisma's implementation. Does that sound like a reasonable to you @dthyresson ?

**Comment by dthyresson** [comment]
> Does that sound like a reasonable to you @dthyresson ?

Yes, but may need to override `migrate reset` as well since it calls their seed command as well.

**Comment by Jolg42** [comment]
Hi @peterp & @dthyresson!

So you could use one of these solutions to make seed execute what you want here:

1 - Add the the package.json a `ts-node` script that `db seed` will check and use if defined
```
"scripts": {
 "ts-node": "ts-node --compiler-options '{\\\"module\\\":\\\"commonjs\\\"}'"
},
```
https://www.prisma.io/docs/reference/api-reference/command-reference#db-seed-preview

I can see how this can be confusing (naming it `ts-node` but it can actually be any command), we'll think about this and see what we can improve.

2 - Creating a `seed.sh` bash script where you could do anything you want.
https://www.prisma.io/docs/guides/database/seed-database#seeding-your-database-with-a-bash-script

**Comment by peterp** [comment]
@Jolg42 Awesome, thanks!

**Comment by peterp** [comment]
@dthyresson What if we provided a default `seed.sh` script which called `yarn babel-node ./seed.ts`? I think that may use our babel configurations and solve most of our problems.

**Review by peterp — COMMENTED** [review]
Based on an earlier discussion: We're either going to leave this as an ordinary JS file, or we're going to try one last thing with the seed.sh file and `yarn babel-node`

**Comment by thedavidprice** [comment]
@dthyresson nudging you on this one — we decided to _not_ support TS, correct? If so, what are the next steps with this one?

**Comment by dthyresson** [comment]
> @dthyresson nudging you on this one — we decided to _not_ support TS, correct? If so, what are the next steps with this one?

@thedavidprice I think based on our other discussions, we said we're ok with this seed file being simply a .js file. 

Therefore, I think we can close out this PR and not merge these changes in.

We can revisit post 1.0 when/if Prisma changes how their seed command works (as is still officially a preview feature).

A good next step would be to work with them to 

1. let the ts version of the seed file run without the necessary
```
    "ts-node": "ts-node --compiler-options '{\"module\":\"CommonJS\"}'"
```

2. maybe let their seed and or db reset command run a user-defined script that Redwood can run it its way

**Comment by thedavidprice** [comment]
@dthyresson Got it. I'll let you do the closing!

Should we add a comment in the template file that TS is not supported... and/or document this somewhere?

**Comment by dthyresson** [comment]
> @dthyresson Got it. I'll let you do the closing!
> 
> Should we add a comment in the template file that TS is not supported... and/or document this somewhere?

@thedavidprice Have added a message in the seed cli command docs here: https://github.com/redwoodjs/redwoodjs.com/pull/754

Closing.

**Pull request closed without merging**